examples/Azure_IoT_Adu_ESP32/Azure_IoT_Adu_ESP32.ino

```diff
@@ -0,0 +1,1024 @@
+
+// Auxiliary functions
+#ifndef IOT_CONFIG_USE_X509_CERT
+static AzIoTSasToken sasToken(
+    &hub_client,
+    AZ_SPAN_FROM_STR(IOT_CONFIG_DEVICE_KEY),
+    AZ_SPAN_FROM_BUFFER(sas_signature_buffer),
+    AZ_SPAN_FROM_BUFFER(mqtt_password));
+#endif // IOT_CONFIG_USE_X509_CERT
+
+static void connect_to_wifi()
+{
+  Logger.Info("Connecting to WIFI SSID " + String(ssid));
+
+  WiFi.mode(WIFI_STA);
+  WiFi.begin(ssid, password);
+  while (WiFi.status() != WL_CONNECTED)
+  {
+    delay(500);
+    Serial.print(".");
+  }
+
+  Serial.println("");
+
+  Logger.Info("WiFi connected, IP address: " + WiFi.localIP().toString());
+}
+
+static void initialize_time()
+{
+  Logger.Info("Setting time using SNTP");
+
+  configTime(GMT_OFFSET_SECS, GMT_OFFSET_SECS_DST, NTP_SERVERS);
+  time_t now = time(NULL);
+  while (now < UNIX_TIME_NOV_13_2017)
+  {
+    delay(500);
+    Serial.print(".");
+    now = time(nullptr);
+  }
+  Serial.println("");
+  Logger.Info("Time initialized!");
+}
+
+// get_request_id sets a request Id into connection_request_id_buffer and
+// monotonically increases the counter for the next MQTT operation.
+static az_span get_request_id(void)
+{
+  az_span remainder;
+  az_span out_span = az_span_create(
+      (uint8_t*)connection_request_id_buffer, sizeof(connection_request_id_buffer));
+
+  connection_request_id++;
+  if (connection_request_id == UINT16_MAX)
+  {
+    // Connection id has looped.  Reset.
+    connection_request_id = 1;
+  }
+
+  az_result rc = az_span_u32toa(out_span, connection_request_id, &remainder);
+  if (az_result_failed(rc))
+  {
+    Logger.Info("Failed to get request id");
+  }
+
+  return az_span_slice(out_span, 0, az_span_size(out_span) - az_span_size(remainder));
+}
+
+static void prvParseAduUrl(az_span xUrl, az_span* pxHost, az_span* pxPath)
+{
+  xUrl = az_span_slice_to_end(xUrl, sizeof("http://") - 1);
+  int32_t lPathPosition = az_span_find(xUrl, AZ_SPAN_FROM_STR("/"));
+  *pxHost = az_span_slice(xUrl, 0, lPathPosition);
+  *pxPath = az_span_slice_to_end(xUrl, lPathPosition);
+}
+
+void update_started() { Serial.println("CALLBACK:  HTTP update process started"); }
+
+void update_finished() { Serial.println("CALLBACK:  HTTP update process finished"); }
+
+void update_progress(int cur, int total)
+{
+  Serial.printf("CALLBACK:  HTTP update process at %d of %d bytes...\n", cur, total);
+}
+
+void update_error(int err) { Serial.printf("CALLBACK:  HTTP update fatal error code %d\n", err); }
+
+static az_result download_and_write_to_flash(void)
+{
+  az_span url_host_span;
+  az_span url_path_span;
+  prvParseAduUrl(adu_update_request.file_urls[0].url, &url_host_span, &url_path_span);
+
+  WiFiClient client;
+
+  httpUpdate.onStart(update_started);
+  httpUpdate.onEnd(update_finished);
+  httpUpdate.onProgress(update_progress);
+  httpUpdate.onError(update_error);
+  httpUpdate.rebootOnUpdate(false);
+
+  // TODO: Add SHA check of image (don't immediately reboot)
```

Comment:
Is this different than verify_image below?

Reply:
ahh yup I added this at `verify_image()`